R/row_policy.R

```diff
@@ -0,0 +1,196 @@
+#' Row policy
+#' 
+#' @export
+#' @inheritParams grant
+#' @param name (character) scalar name for the policy. required
+#' @examplesIf has_postgres()
+#' library(DBI)
+#' library(RPostgres)
+#' con <- dbConnect(Postgres())
+#' if (!dbExistsTable(con, "passwd")) {
+#'    setup_example_table(con, "passwd")
+#' }
+#' rls_tbl(con, "passwd") %>%
+#'   row_policy("my_policy") %>%
+#'   rls_run()
+#' rls_policies(con)
+#' 
+#' # cleanup
+#' rls_drop_policies(con)
+#' dbDisconnect(con)
+row_policy <- function(.data, name) {
+  pipe_autoexec(toggle = rls_env$auto_pipe)
+  assert_is(name, "character")
+  assert_scalar(name)
+  .data <- as_row_policy(.data)
+  .data$name <- name
+  .data
+}
+
+#' Commands
+#' 
+#' @export
+#' @inheritParams grant
+#' @examplesIf has_postgres()
+#' library(DBI)
+#' library(RPostgres)
+#' con <- dbConnect(Postgres())
+#' if (!dbExistsTable(con, "passwd")) {
+#'    setup_example_table(con, "passwd")
+#' }
+#' rls_tbl(con, "passwd") %>%
+#'   row_policy("their_policy") %>%
+#'   commands(update)
+#' 
+#' # cleanup
+#' dbDisconnect(con)
+commands <- function(.data, ...) {
+  pipe_autoexec(toggle = rls_env$auto_pipe)
+  .data <- as_row_policy(.data)
+  .data$commands <- dot_names(...)
+  .data
+}
+
+#' Create rule for existing rows
+#' 
+#' @export
+#' @inheritParams grant
+#' @param using an expression to use to check against existing rows
+#' @param sql (character) sql syntax to use for existing rows
+#' @details Use either `using` or `sql`, not both
+#' @examplesIf has_postgres()
+#' library(DBI)
+#' library(RPostgres)
+#' con <- dbConnect(Postgres())
+#' if (!dbExistsTable(con, "passwd")) {
+#'    setup_example_table(con, "passwd")
+#' }
+#' rls_tbl(con, "passwd") %>%
+#'   row_policy("a_good_policy") %>%
+#'   commands(update) %>%
+#'   rows_existing(sql = 'current_user = "user_name"')
+#' 
+#' # cleanup
+#' dbDisconnect(con)
```

Comment:
Can you also write an example where you would use `using`? I think I personally need it to wrap my head around this.

Reply:
Yes, will do

Reply:
so i think it's best to drop NSE for `rows_existing` and `rows_new` (which I did) - so not adding egs as youd expect